test/collection/test_validator.py

```diff
@@ -0,0 +1,159 @@
+from dataclasses import dataclass
```

Comment:
I do not like these tests :/ 
They are hard to read, needlessly complicated and the mocks will probably break regularly.

May I suggest simply doing the call eg `client.collections.exists(1)` with pytest.raises?

Reply:
The mocks are only there so that the methods can be called from the parent object in which they reside. The mock does nothing (except pretend to know a server version) and the tests only test the initial validation logic in each function.

The complication is to reduce the amount of repetitive writing of tests and minimise duplicate code. The test at the bottom still does:
```python
def test_validator(test: TestableFunction):
    with pytest.raises(WeaviateInvalidInputException) as e:
        test.func()
    assert e.value.args[0] == f"Invalid input provided: {test.err}."
```
It's just been parametrised over all the testable functions

Reply:
>The complication is to reduce the amount of repetitive writing of tests and minimise duplicate code.

It is 160 lines, might not repetetive but it is very hard to read. You must follow strings to some dictionary and then look at list indices with indirect fixtures.

How about this?
```
@pytest.fixture(scope="module")
def col() -> Collection:
    client = weaviate.connect_to_local()
    client.collections.delete_all()
    col = client.collections.create("Test")
    yield col
    client.collections.delete_all()


@pytest.mark.parametrize("parameter,error", [(1, "<class 'int'>"), ("who", typing.List[str])])
def test_tenant_update(col: Collection, parameter: Union[int, str], error: str) -> None:
    with pytest.raises(WeaviateInvalidInputException) as e:
        col.tenants.update([parameter])  # type:ignore
        assert (
            e.value.args[0]
            == f"Invalid input provided: Argument 'tenants' must be typing.List[weaviate.collections.classes.tenants.Tenant], but got {error}."
        )
```

it is 8 lines for a self-contained, parametrized test that is very easy to parse and even shorter than than just the definition in your `functions`-fixture

Reply:
> is key to these being self-contained unit tests rather than integration tests

Does this matter? You're getting clearly worse and harder to read/understand code for the benefit of following some arbitrary "rule". What matters for me is that it is actually tested, calling them unit or integration tests is just semantics

Reply:
Or to formulate it differently: What are the benefits of having unit instead of integration tests? On the downside we have the more complicated tests, what are the upsides?

Reply:
The benefit of these being unit tests is that they run without needing a connection to Weaviate. Since they don't test anything to do with the connection to Weaviate, they only test that the functions throw exceptions immediately, they don't have a connection to Weaviate (hence the mock)

Reply:
Will remove this file from this PR and we can revisit unit tests for validation later